ASCII codes

In Binary Strings:                     
        
    a   | 01100001   A  01000001   spacebar  00100000   ?   00111111
    b   | 01100010   B  01000010   +         00101011   |   01111100
    c   | 01100011   C  01000011   -         00101101
    d   | 01100100   D  01000100   *         00101010
    e   | 01100101   E  01000101   /         00101111
    f   | 01100110   F  01000110   .         00101110
    g   | 01100111   G  01000111   0         00110000
    h   | 01101000   H  01001000   1         00110001
    i   | 01101001   I  01001001   2         00110010
    j   | 01101010   J  01001010   3         00110011
    k   | 01101011   K  01001011   4         00110100
    l   | 01101100   L  01001100   5         00110101
    m   | 01101101   M  01001101   6         00110110
    n   | 01101110   N  01001110   7         00110111
    o   | 01101111   O  01001111   8         00111000
    p   | 01110000   P  01010000   9         00111001
    q   | 01110001   Q  01010001   !         00100001
    r   | 01110010   R  01010010   @         01000000
    s   | 01110011   S  01010011   #         00100011
    t   | 01110100   T  01010100   $         00100100
    u   | 01110101   U  01010101   %         00100101
    v   | 01110110   V  01010110   ^         01011110
    w   | 01110111   W  01010111   <         00111100
    x   | 01111000   X  01011000   >         00111110
    y   | 01111001   Y  01011001   &         00100110
    z   | 01111010   Z  01011010   ,         00101100

In Hexadecimal Strings:                     
        
    a   61   A  41   spacebar 20    ?   3f
    b   62   B  42   +         2b   |   7c
    c   63   C  43   -         2d
    d   64   D  44   *         2a
    e   65   E  45   /         2f
    f   66   F  46   .         2e
    g   67   G  47   0         30
    h   68   H  48   1         31
    i   69   I  49   2         32
    j   6a   J  4a   3         33
    k   6b   K  4b   4         34
    l   6c   L  4c   5         35
    m   6d   M  4d   6         36
    n   6e   N  4e   7         37
    o   6f   O  4f   8         38
    p   70   P  50   9         39
    q   71   Q  51   !         21
    r   72   R  52   @         40
    s   73   S  53   #         23
    t   74   T  54   $         24
    u   75   U  55   %         25
    v   76   V  56   ^         5e
    w   77   W  57   <         3c
    x   78   X  58   >         3e
    y   79   Y  59   &         26
    z   7a   Z  5a   ,         2c

For Alphabetic Strings. if we want to encode a small letter, it encode that by its corresponding capital letter, & for capital letter remains same whereas if we input any other symbols, it does not give anything.

**Shift Cipher**

For this cryptosystem $\mathcal{{P} = {C} = {K}} = {\mathbb{Z}_{26}}$.  So, the possible keys $k$ are $0 \le k \le 25 k$. 
* Encryption rule is $e_{k}(x) = (x + k)$ mod $26$.

* Decrytion rule is $d_{k}(y) = (y-k)$ mod $26$

  where $x$ is the plaintext & $y$ is ciphertext.
  
**In Particular**, for $k = 3$, we called the cryptosystem as Casear cipher.

In [1]:
S = ShiftCryptosystem(AlphabeticStrings())
S
# S.key_space() #by this we can find key_space() of the corresponding cryptosystem

#S.alphabet_size() # this will tells us how many elements are there in the corresponding cryptosystem
P = S.encoding("we will meet at midnight"); # this is just encoding the string monoid over the shift cipher S 
P
#K = 11 #secret key must be lies between 0 to 25
key = S.random_key()  #, using this command we can also generate  random key within this cipher key space.This can be done for every cryptosystem.
C = S.enciphering(key,P) #encrypting the given plaintext
C
D = S.deciphering(key,C); #decrypt the plaintext
D

WEWILLMEETATMIDNIGHT

Remind that, don't give a space between letters, except AlphabeticStrings(), every strings take a value due to this space.
In hexadecimal , encoding value of one space is **20**.

In [2]:
S = ShiftCryptosystem(HexadecimalStrings())
S
# S.key_space() #by this we can find key_space() of the corresponding cryptosystem
P = S.encoding("wewillmeetatmidnight"); # this is just encoding the string monoid over the shift cipher S 
P
K = 11 #secret key must be lies between 0 to 15
C = S.enciphering(K,P) #encrypting the given plaintext
C
D = S.deciphering(K,C); #decrypt the plaintext
D
D == P

True

In [3]:
S = ShiftCryptosystem(BinaryStrings())
S
# S.key_space() #by this we can find key_space() of the corresponding cryptosystem
# S.alphabet_size() # this will tells us how many elements are there in the corresponding cryptosystem
P = S.encoding("wewillmeetatmidnight"); # this is just encoding the string monoid over the shift cipher S 
P
K = 1 #secret key should be either 0 or 1
C = S.enciphering(K,P) #encrypting the given plaintext
C
D = S.deciphering(K,C); #decrypt the plaintext
D
P == D

True

In [4]:
P

0111011101100101011101110110100101101100011011000110110101100101011001010111010001100001011101000110110101101001011001000110111001101001011001110110100001110100

In [5]:
d=  S.brute_force(C)

In [6]:
for i in range(0,len(list(d.values()))):
    if list(d.values())[i] == P:
        print(True)
        print(i)

True
1


**Substitution Cipher**: 

For this cryptosystem $\mathcal{{P} = {C}} = {\mathbb{Z}_{26}}$. The keyspace $\mathcal{K}$ is consists of all possible permutation of the $26$ symbols $0,1,\cdots,25$, so $\mathcal{K} = {26!}$.For each permutation $\pi \to \mathcal{K}$,

* Encryption rule : $e_{\pi}(x) = \pi(x)$.

* Decryption rule : $d_{\pi}(y) = {\pi}^{-1}(y)$, where ${\pi}^{-1}$ is the inverse permutation to $\pi$ with $x$ as plaintext & $y$ as ciphertext.



In [7]:
S = SubstitutionCryptosystem(AlphabeticStrings())
S
P = S.encoding("wewillmeetatmidnight"); # this is just encoding the string monoid over the shift cipher S 
P
K = S.random_key() # here generate a random key
K
C = S.enciphering(K,P) #encrypting the given plaintext
C
D = S.deciphering(K,C); #decrypt the plaintext
D

WEWILLMEETATMIDNIGHT

In [8]:
S = SubstitutionCryptosystem(BinaryStrings())
S
P = S.encoding("wewillmeetatmidnight"); # this is just encoding the string monoid over the shift cipher S 
P
K = S.random_key() 
K
C = S.enciphering(K,P) #encrypting the given plaintext
C
D = S.deciphering(K,C); #decrypt the plaintext
D
D == P

True

In [9]:
S = SubstitutionCryptosystem(HexadecimalStrings())
show(S)

#S.alphabet_size() # this will tells us how many elements are there in the corresponding cryptosystem
P = S.encoding("wewillmeetatmidnight"); # this is just encoding the string monoid over the shift cipher S 
show(P)
K = S.random_key() 
show(K)
C = S.enciphering(K,P) #encrypting the given plaintext
show(C)
D = S.deciphering(K,C); #decrypt the plaintext
show(D)
D == P

Substitution cryptosystem on Free hexadecimal string monoid

776577696c6c6d65657461746d69646e69676874

8796240fb3e1c5da

ff04ff030c0c050404f207f20503020d030f0bf2

776577696c6c6d65657461746d69646e69676874

True

**Affine cipher** :

For this cryptosystem $\mathcal{{P} = {C}} = \mathbb{Z}_{26}$. The keyspace $\mathcal{K}$ is $[(a,b) \to \mathbb{Z}_{26} \times \mathbb{Z}_{26} : gcd(a,26) = 1]$. For key $k = (a,b) \to \mathcal{K}$, 

* Encryption rule is : $e_{k}(x) = (ax +b)$ mod $26$, $x$ is plaintext.

* Decryption rule is : $d_{k}(x) = a^{-1}(y -b)$ mod $26$, $y$ is ciphertext.

**Note:** The number of keys in affine cipher over $Z_{m}$ is $m.{\phi(m)}$.

In [10]:
S = AffineCryptosystem(AlphabeticStrings())
S

#S.alphabet_size() # this will tells us how many elements are there in the corresponding cryptosystem
P = S.encoding("Iwillcallyouback"); # this is just encoding the string monoid over the shift cipher S 
show(P)
# K = S.random_key() 
K = (17,11) #gcd(a = K[0],26) should be  1
show(K)
C = S.enciphering(K[0],K[1],P) #encrypting the given plaintext
show(C)
D = S.deciphering(K[0],K[1],C); #decrypt the plaintext
show(D)
D == P

IWILLCALLYOUBACK

(17, 11)

RVRQQTLQQDPNCLTZ

IWILLCALLYOUBACK

True

**Vigener cipher**:

In this cryptosystem $\mathcal{{P} = {C} = {K}}= {(\mathbb{Z}_{26})}^m$ for  a $m$ positive integer. For a key $k = (k_1,k_2, \cdots, k_{m})$,
* Encryption rule : $e_{k}(x_1,x_2,\cdots, x_{m}) = (x_1+k_1,x_2+k_2,\cdots, x_{m} + k_{m})$, $x$ is plaintext and 

* Decryption rule : $d_{k}(y_1,y_2,\cdots,y_{m}) = (y_1 -k_1,y_2 - k_2, \cdots, y_{m} - k_{m})$,$y$  is ciphertext,

all these operations are performed in $Z_{26}$.

**Note**: This is a polyalphabetic cryptosystem.

In [11]:
V = AlphabeticStrings()
S = VigenereCryptosystem(V,6)
P = S.encoding("thiscryptosystemisnotsecure")

#K = S.random_key()

K = V('CIPHER')
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [12]:
V = HexadecimalStrings()
S = VigenereCryptosystem(V,6)
P = S.encoding("thiscryptosystemisnotsecure")
# S.alphabet_size() # this will tells us how many elements are there in the corresponding cryptosystem
#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
K = S.random_key()
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [13]:
V = BinaryStrings()
S = VigenereCryptosystem(V,6)
P = S.encoding("thiscryptosystemisnotsecure")

#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
K = S.random_key()
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [14]:
ki = S.inverse_key(K)
e = S(K)
d = S(ki)
e(d(V("100001"))) == V("100001") #e,d can interchange
#this tells us that if we encode the string through the 
#given(alphabetic/binary/hexadecimal/octal/radix) Strings(),it will be same 
# with left hand, as decryption is the inverse of encryption rule.

True

**Hill Cipher**:

In this cryptosystem $\mathcal{{P} = {C}} = {(\mathbb{Z}_{26})}^m$ for $m \ge 2$ integer. The keyspace is  $\mathcal{K}$ = $m \times m$ invertible matrices over $\mathbb{Z}_{26}$. Then, for a key $K \to \mathcal{K}$, 

* Encryption rule : $e_{k}(x) = xK$, $x$ is plaintext,

* Decryption rule : $d_{k}(x) = yK^{-1}$, $y$ is ciphertext,

where all operations are performed in $Z_{26}$.

In [15]:
V = AlphabeticStrings()
S = HillCryptosystem(V,6)
P = S.encoding("good da") #when we encode some string, need to care that,
# the string should be multiple of the positive integer m =6,here


#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
K = S.random_key()
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [16]:
V = AlphabeticStrings()
S = HillCryptosystem(V,3)
P = S.encoding("Good day,mate! How was the tour?")

#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
R = IntegerModRing(26) #26 is as we do all operation under Z_26
M = MatrixSpace(R,3,3) #3,3 is the order of the matrix

K = M([[0,3,89],[2,5,20],[3,13,31]]) #remind that this matrix has to be invertible.
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [17]:
V = BinaryStrings()
S = HillCryptosystem(V,3)
P = S.encoding("101")

#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
R = IntegerModRing(2) #26 is as we do all operation under Z_26
M = MatrixSpace(R,3,3) #3,3 is the order of the matrix

K = M([[0,3,89],[2,5,20],[3,13,31]]) #remind that this matrix has to be invertible.
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
#D
D == P

True

In [18]:
V = HexadecimalStrings()
S = HillCryptosystem(V,3)
P = S.encoding("abc")

#K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
R = IntegerModRing(16) #26 is as we do all operation under Z_26, 
# so when we change the strings, like when we give hexadecimal, 
#then we need to give it 16,for binary need to give 2 instead of 26.
M = MatrixSpace(R,3,3) #3,3 is the order of the matrix

K = M([[0,3,6],[2,5,27],[89,57,31]])
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D

D == P

True

**Permutation(Transposition) cipher**:

In this cryptosystem $\mathcal{{P} = {C}} = (\mathbb{Z}_{26})^m$, for some positive $m \to \mathbb{Z}$. The keyspace is $\mathcal{K} = m!$. For a key i.e. a permutation  $\pi \to (\mathbb{Z}_{26})^m$,

* Encryption rule : $e_{\pi}(x_1, \cdots, x_{m}) = (x_{{\pi}_1}, \cdots, x_{{\pi}_m}$, $x$ is the plaintext.

* Decryption rule : $d_{\pi}(y_1, \cdots, y_m) = (y_{{\pi}^{-1}_1},\cdots, y_{{\pi}^{-1}_m})$, $y$ is the ciphertext

where ${\pi}^{-1}$ is the inverse permutation  of $\pi$.

In [19]:
V = AlphabeticStrings()
m = 3
S = TranspositionCryptosystem(V,m)
P = S.encoding("ICT") #given string must be multiple of m

K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
# K = [m - i for i in  range(m)] #using this we can create our own key
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
# for checking ,D == P

ICT

In [20]:
V = HexadecimalStrings()
m = 3
S = TranspositionCryptosystem(V,m)
P = S.encoding("Goodda") #given string must be multiple of m

K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
#K = [(m - i for i in  range(m))] #using this we can create our own key
show(K)
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

(1,2,3)

True

In [21]:
V = BinaryStrings()
m = 3
S = TranspositionCryptosystem(V,m)
P = S.encoding("Goodda") #given string must be multiple of m

K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
# K = [m - i for i in  range(m)] #using this we can create our own key
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

In [22]:
V = OctalStrings()
m = 3
S = TranspositionCryptosystem(V,m)
M = "257"
P = V(M) #given string must be multiple of m

K = S.random_key()
#since in hexadecimal, there are only 16 alphabets, then always key will be 
# combination of 0 to 9 & a to f.
# K = [m - i for i in  range(m)] #using this we can create our own key
K
C = S.enciphering(K,P)
C
D = S.deciphering(K,C)
D
D == P

True

#### RSA algorithm in for encryption and decryption.

$1$. Choose two primes p and q and let n = pq.

$2$.Let e $\in \mathbb{Z}$ be positive such that gcd($e,\phi(n)$) = 1.

$3$.Let d $\in \mathbb{Z}$ be the multiplicative inverse of $e$ modulo $\phi(n)$.

$4$. Our public key is the pair $(n,e)$ and our private key is the pair $(n,d)$.

$5$.For any non-zero integer $m <n$, encrypt $m$ using $c = m^{e}$ (mod $n$).

$6$. Decrypt $c$ using $m = c^{d}$ (mod $n)$.

In [23]:
p = next_prime(randint(10^20,10^30))
show(p)
q = next_prime(randint(10^20,10^39))
show(q)
is_prime(p);is_prime(q)
n = p*q
show(n)
e = ZZ.random_element(1,euler_phi(n))
show(e)

35721808462523611266646549591

555347389800294322712192994700359927589

19838013088608552377353161733890680209415024601895127496409457566099

4992519480740022193138685216579876779103516689633225213699120325768

ZZ.random_element()
*  if no input is given , it returns one random integer.
* if one input is given , it return an integer between 0 to n-1 

In [24]:
while gcd(e,euler_phi(n)) != 1:
    e = ZZ.random_element(euler_phi(n))

In [25]:
e

18265420309859201265343005928204050226741875159077218129492083340709

In [26]:
bool(e <n)

True

In [27]:
d = inverse_mod(e,euler_phi(n))
d

5739187852539587518585478151912079451178635827020130280157366658829

In [28]:
mod(d*e,euler_phi(n))  ==1

True

In [29]:
m = ZZ.random_element(n) #plaintext
m

220493635236657893887831846356762493387911039332383252793893225496

In [30]:
m < n

True

In [31]:
ciphertext = power_mod(m,e,n)
ciphertext

16543609564518855078122784340302011768232781276644193310078505746474

In [32]:
power_mod(ciphertext,d,n) == m

True

# Primality Testing

### Euler's Criterion for Quadratic Residues

Let p be an odd prime. Then $a$ is quadratic residue modulo p if and only if $a^{\frac{p-1}{2}} \equiv 1$ (mod p).

In [33]:
p = 293
p.is_prime()
 

True

In [34]:
a = 45482
(a**((p-1)/2))%p
# so a is quadratic residue modulo p

1

In [35]:
a1 = 33354653
(a1**((p-1)/2))%p
# so a1 is not a quadratic rsidue modulo p

292

### Legendre Symbol $\left(\frac{a}{p}\right)$

Suppose  p is an odd prime. For any integer $a$, define the Legendre Symbol $\left(\frac{a}{p}\right)$ as :

$\left(\frac{a}{p}\right) = \begin{cases}
0,~~~~~~~ \mbox{a $\equiv$ 0 modulo p}\\
1, ~~~~~~~  \textrm{a is quadratic residue modulo p}\\
-1,~~~~~~  \mbox{a is quadratic non-residue modulo p}
\end{cases}$



In [36]:
p = 2847
p.is_prime()

False

In [37]:
p = p.next_prime()
show(p)
a = 25
legendre_symbol(a,p)

2851

1

In [38]:
a = 790
legendre_symbol(a,p)

-1

In [39]:
a = 2*p
legendre_symbol(a,p)


0

In [40]:
# it shows that p has to be a odd prime
legendre_symbol(245,56)

ValueError: p must be a prime

### Solovay-Strassen test

**Algorithm :**

* input is a positive integer n.

* Take a random number $1 \le a \le n-1$.

* If (a,n) > 1, then n is composite.

* Compute $a^{\frac{n-1}{2}},\left(\frac{a}{p}\right)$.

* Check if $a^{\frac{n-1}{2}}\not= \left(\frac{a}{p}\right)$(mod n).Then $n$ is composite. 

* else choose another $a$ and repeat the process.



In [43]:
import random

In [44]:
def solovay_strassen(n):
    a = random.randint(1,n-1)
    x = kronecker(a,n)
    if (x==0):
        print(f"{n} is composite")
    s = ((n-1)/2)
    y = (a**s) % n
    if (x==y):
        print(f"{n} prime")
    else:
         print(f"{n} is composite")

In [45]:
solovay_strassen(293)

293 is composite


###  Miller-Rabin  test

If we have factorization for $n-1$, then this test wil work. 

Goal is to check $n$ is prime or not.

Suppose $n-1 = 2^{k}m, k\ge 1,m$ is odd.
Choose a random integer $a, 1 \le a \le n-1$.

Let $b = a^{m}$(mod $n$). If $b \equiv 1$(mod $n)$, then $n$ is prime.

Take $i = 1$, check if $b \equiv -1$(mod $n)$, then $n$ is prime. 

Otherwise $b = b^2$(mod $n)$ and increase $i$ by 1 and repeat the process until $i$ reach $k$.

In [46]:
# Miller- Rabin test

def miller_rabin(n):
    if (n <=3) :
        raise Exception('n should be greater than 3')
    if (n%2 == 0):
        return False
    # find u odd such that n-1 = 2^k*u
    u = n-1
    k =0
    while(u %2 ==0):
        u //= 2
        k += 1
    a = random.randint(1,n-1) #select a randomly
    b = pow(a,u,n)
    if (b ==1 or b==n-1): #b==-1 is same as n-1
        return True
    for j in range(1,k):
        b = (b*b)%n
        if b == n-1: return True
        if b ==1:
            return False
    return False

In [47]:
miller_rabin(3154050820)

False

In [48]:
miller_rabin(40530101879)

True

In [49]:
3154050820.is_prime()

False

In [50]:
40530101879.is_prime()

True

## Factorization

**Pollard P-1 method :**

In this algorithm has two inputs : the integer $n$ to be factorized and a prespecified bound B. 

$1$. Set $a = 2$, generally we choose this, another convenient $a$.

$2$. Take $j = 2,3,\cdots, B$ and compute $a =a^{j}$(mod $n$).

$3$. compute $d = (a-1,n)$ where $(,)$ denotes gcd of two numbers.

$4$. If $1 < d< n$, then $d$ is a divisor of $n$ and return $d$.

$5$. Otherwise keep on increasing $j$ upto B. Otherwise we can change the value of $a$.

Once we get one factor $d$ of $n$, easily we can compute other factor by $\frac{n}{d}$.

In [51]:
def Pollard_algo(n,B):
    a = 2  # generally we choose a = 2
    for j in range(2,B,1):
        a = (a**j)%n
        d = gcd(a-1,n)
        if (1 < d < n) and (d.is_prime() == True):
            pass
    print(d)

In [52]:
timeit('Pollard_algo(15770708441,180)')

625 loops, best of 3: 340 μs per loop

In [53]:
factor(135979-1)

2 * 3 * 131 * 173

In [54]:
d1 = Pollard_algo(15770708441,180) #example is taken from book
d1

135979



since d1-1 has highest factor 173, so for any integer greater than 173, d1-1 will divide the factorial(B).

In [55]:
(135979-1)//(factorial(180))

0

In [56]:
d2 = 15770708441//135979
d2

115979

So, $d1,d2$ are two factors of given $n = 15770708441$.

In the Pollard $p-1$ method, there are $B-1$ modular exponentiations, each requiring at most $2log_{2}B$ modular multiplications using the 'square and multiply algorithm'. The gcd can be computed in time $O((logn)^3)$ using the 'extended euclidean algorithm.Hence the complexity of the algorithm is $O(BlogB(logn)^2 + (logn)^3)$.  

If the integer $O((logn)^i)$ for a fixed integer $i$, then the algorithm isindeed a polynomial-time algorithm(as a function of $log(n)$). 

However, for such a choice of $B$, the probability of success will be very small. On the other hand, if we increase the value of $B$,then this algorithm guaranteed to be successful but it's speed will be decreased.

The drawback of this method is that it requires a prime factor $p$ such that $p-1$ has only small prime factors.

**Pollard Rho method :**

We want to factor a integer $n$.

Let S be a finite  set and take a function $f : S \to S$. Choose $s_{0} \in S$, compute $s_{i+1} = f(s_{i})  \forall i = 1,2,\cdots$.

Store those values of the sequence in a list. Since S is finite, then $s_{i} \equiv s_{j}$(mod $n)$. Using the fact that, $f$ is polynomial function of integer coefficients, then we get $f(s_{i}) \equiv f(s_{j})$(mod $n$). so,  $s_{i+1}=s_{i+2}$(mod $n)$.
So $\exists, s_{i} \equiv s_{2i}$(mod $n)$. Then $(s_{i}-s_{2i},n) = d$, is the non-trivial factor of $n$.

In [57]:
def f(x):
    return x**2  + x +1


def Pollard_rho(n):
    seed = 2
    X = [seed,f(seed)%n] #1st is s0, 2nd one is s1
    X.append(f(X[1])%n) #append s2
    k = 1
    while gcd(X[2*k] - X[k],n) ==1:
        for j in range(2):
            X.append(f(X[2*k +j])%n)
        k += 1
    outfact = gcd(X[2*k] - X[k],n)
    print([outfact,n/outfact])

In [58]:
Pollard_rho(455459)

[613, 743]


## Elgamal Cryptosystem

Suppose Alice  wants to send a message $x$ to Bob. 

From Bob's side there will be the key generation algorithm. 

**key generation algorithm :**

* Choose a large prime p.

* Choose a generator or primitive root $\alpha$.

* Choose $d \in \{2,3,\cdots,p-2 \}$, compute $\beta = {\alpha}^d ($mod$p)$. 

Now Alice will choose $i \in \{2,3,\cdots,p-2\}$ and then compute $k_{e} = {\alpha}^{i}($mod$p)$. After that, compute the massking key $k_{m} = {\beta}^{i}($mod$p)$. 

**Encryption function :**

$E(x) = x \times {\alpha}^{i}($mod$p) = y$,say. The ciphertext is the orderd-pair $(k_{e},y)$ to be send to Bob from Alice. 

**Decryption process :**

For decryption first Bob will compute $k_{m} = {(k_{e})}^{d}$. Then $D(x) = (y.(k_{m})^{-1})$(mod $p)$, that should be equal to $x$.

In [59]:
import random;
p = 29;alpha = 2;
l = list(range(2,p-2,1)) # create the list by  the elements 2 ,3,..29
d = l[random.randrange(len(l))] # selecting the random element
beta = pow(alpha,d,p)
show(beta)
show(d)
x  =l[random.randrange(len(l))] #26 # message is also belong to Z_29*

i = l[random.randrange(len(l))]

k_e = power_mod(alpha,i,p)

show(k_e)

print(x)

massking_key = power_mod(beta,i,p)
show(massking_key)

encryption = (x*pow(alpha,d*i,p))%p
ciphertext = (massking_key,encryption)


show(encryption)

km_inv = pow(massking_key,-1,p)
show(km_inv)

print(pow(massking_key*km_inv,1,p))

decryption = pow(encryption*km_inv,1,p)
show(decryption)


decryption == x


19

9

6

26


22

21

4

1


26

True

### Baby steps, Giant steps

Algorithm of Baby-Steps,Giant steps for solving discrete logarithm problem :

that is we need to find $x$ such that $y = g^{x}$(mod $p)$ for an odd prime $p$. Let $m = \lceil x \rceil$.

So, now $y = g^{x} = g^{im}\circ g^{j}$(mod $p)$ imples that $y.g^{-i.m} = g^{j}$(mod $p$).

In Baby steps, we compute $g^{j}$ for $j \in \{0,1,\cdots m-1\}$.Store it in a list,named X.

In Giant steps, we compute $y.g^{-i.m}$ for $i \in \{0,1,\cdots m-1\}$.Store it in an another list,named Y.

Then look for those indices $i_{0},j_{0}$ of X and Y respectively where $X[i_{0}]$ is same as $Y[j_{0}]$. Then our desired $x$ will be $x = i + j\times m$



In [60]:
def bs_gs(p,alpha,beta):
    m = ceil(sqrt(p-1))
    inv_m = power_mod(alpha,-1,p)
    X = []
    # baby steps
    for j in range(m):
        X.append(pow(alpha,j,p))
    # giant steps
    Y=[]
    for i in range(m):
        y = (22*pow(inv_m,i*m,p))%p
        Y.append(y)    
    X_set=set(X)  
    Y_set=set(Y)
    common=list(X_set.intersection(Y_set))
    temp_X=0
    for i in range(len(X)):
        if X[i]==common[0]:
            temp_X=i
            break
        else:
            pass
    
    temp_Y=0
    for i in range(len(Y)):
        if Y[i]==common[0]:
            temp_Y=i
            break
        else:
            pass
    
    return (temp_X + m*temp_Y)%(p-1)

In [61]:
bs_gs(37,2,5)

31

### Day  2 of seminar

### Pollard Rho method for DLP

In **Pollard Rho method** for **DLP(Discrete Logarithm Problem)** , First we partitioned the Group into three sets $S_{0},S_1,S_2$ of roughly equal size.

For example, $1 \notin S_1$, then by choosing $x_0 = 1$, we define a sequence of group elements $x_0,x_1,x_2,\cdots$ by $x_{i+1} = 
\begin{cases}
\alpha.x_{i} ~~~~~, x_{i} \in S_0\\
{x_{i}}^2 ~~~~~~~~, x_{i} \in S_1\\
\beta.x_{i} ~~~~~~, x_{i} \in S_2
\end{cases}$ for $i  \ge 0$.

This sequence of group elements $x_0,x_1,x_2,\cdots$ turn out to define two more sequences of integers $a_0,a_1,\cdots$ and $b_0,b_1,b_2,\cdots$ satisfying $x_{i} = {\alpha}^{a_{i}}{\beta}^{b_{i}}$ for $i \ge 0$.  So, $a_0 = 0, b_0 = 0$,  for $i \ge 0$,
$a_{i+1} = 
\begin{cases}
a_{i} +1 ~~~~~~, a_{i} \in S_0\\
2{a_{i}} ~~~~~~, a_{i} \in S_1\\
a_{i} ~~~~~~~~~, a_{i} \in S_2
\end{cases}$
and 
$b_{i+1} = 
\begin{cases}
b_{i} , b_{i} \in S_0\\
2{b_{i}} , b_{i} \in S_1\\
b_{i}+1 , x_{i} \in S_2
\end{cases}$.

We continue finding the finding the tripets $(x_i,a_i,b_i)$ and $x_{2i},a_{2i},b_{2i}$ until $x_i = x_{2i}$. After ending this process,let's take $g_1 = a_i - a_{2i}$ mod$(p-1)$ and $h_1 = b_{2i} - b_{i}$(mod $p-1)$.

If $h_1 \equiv 0$( mod $p-1)$, then the algorithm fails. Let it is not. Then compute $d = \text{gcd}(h1,p-1)$ and also find $s,t$ such that $d = s.h_1 + t(p-1)$.

Now, find that particular $w$ for which $g^{x} \equiv h$(mod $p$) and $x = \frac{su  + (p-1)w}{d}$mod ($p-1)$.

In [62]:
def Pollard_rho_dlp(g,h,p):
    S0 = [k for k in [0..p//3]] #define the three set s0
    S1 = [k for k in [p//3+1..2*p//3]]  #define  the set s1
    S2 = [k for k in [2*p//3+1..p-1]]  # define the set s2
    T1 = [1,0,0] #x0,a0,b0 i.e x_(i-1),a_(i-1),b_(i-1)
    T2 = [1,0,0] # x_(2i-2),a_(2i-2),b_(2i-2)
    def w(t):
        if t[0] in S0: #t is the values  of x,a,b
            return [(g*t[0])%p,(t[1]+1)%(p-1),t[2]] #(alpha.x,a +1 mod p-1,b)
        elif t[0] in S1:
            return [t[0]^2,(2*t[1])%(p-1),(2*t[2])%(p-1)] #(x^2,2a mod p-1,2b mod p-1)
        else:
             return [(h*t[0])%p,t[1],(t[2]+1)%(p-1)] #(beta.x, a,b+1 mod p-1)
    
    jo = True
    while jo:
        T1 = w(T1)
        T2 = w(w(T2))
        if T1[0] == T2[0]:
            jo = False
    
    
    g1 = (T1[1] - T2[1])%(p-1) # (a_i - a_2i) mod (p-1)
    h1 = (T2[2]-T1[2])%(p-1) # (b_2i - b_i) mod (p-1)
    d,s,t = xgcd(h1,p-1)
    possibly = [((s*g1 + i*(p-1))/d)%(p-1) for i in [0..d-1]] #list of possible solutons
    sol0= [j for j in possibly if (g^j)%p == h]
    return sol0[0]
    
        

In [63]:
Pollard_rho_dlp(3,200,283)

97

In [64]:
Pollard_rho_dlp(2,5,1019)

10

### Index calculas method

The Index calculas algorithm is used to solve the discrete logarithm problem $a^x \equiv b$( mod$p)$. First we fix a factor base $B = \{p_1,p_2,\cdots,p_{k}\}$ containing small primes. Consider a sequence $x_1,x_2,\cdots$ of distinct primes from the set $\{0,1,\cdots,p-2\}$. 

Compute $a^{x_{i}}$(mod $p$).and check if the resulting value is only divisible by primes from the set $B$ i.e. $a^{x_{i}} \equiv {p_1}^{u_{i1}}{p_2}^{u_{i2}}\cdots{p_{k}}^{u_{ik}}$(mod $p$).If this is found, then we take `log` in both sides of the above expression,we get $x_{i} \equiv u_{i1}log_{a}(p_1) + u_{i2}log_{a}(p_2) + \cdots + u_{ik}log_{a}(p_k)$mod $(p-1)$.

This gives a linear equation  in the unknows $log_{a}(p_1),\cdots log_{a}(p_k)$. Here we have $k$ many unknows. So, we need to solve  the linear equations.

Finally, we select random integers $m \in \{0,1,\cdots,p-2\}$ and compute $a^{m}b$ mod ($p)$, if it is divisible by primes in $B$, then $log_{a}(b) + m \equiv m_{i1}log_{a}(p_1) + m_{i2}log_{a}(p_2) + \cdots + m_{ik}log_{a}(p_{k})$ mod ($p-1)$. Hence,

$log_{a}(b)  \equiv m_{i1}log_{a}(p_1) + m_{i2}log_{a}(p_2) + \cdots + m_{ik}log_{a}(p_{k}) - m$ mod ($p-1)$.

In [65]:
# B is the factor base
def Index_calculas(a,b,p,B): # find the integer n such that b = a^n
    S = Set([0..p-2]) #define the set 0,1,p-2
    pd = prime_divisors(p-1) #it return the list of prime divisors
    T = True
    M = []
    V =[]
    dimM = 0
    while T:
        xi = S.random_element()
        S = S.difference(Set([xi]))
        q = [valuation((a^xi)%p,k) for k in B] #valuation(b,a) is compute for which integer n such that  a^n =b
        if prod([k^valuation((a^xi)%p,k) for k in B]) == (a^xi)%p:
            M0 = M + [q]
            
            if all(Matrix(GF(x),M0).rank() == dimM+1 for x in pd):
                M.append(q)
                V.append(xi)
                dimM = dimM +1
            else:
                pass
        if dimM == len(B):
            T = False
    LOGs = matrix(Integers(p-1),M).inverse()*vector(V)
    S = Set([0..p-2])
    T = True
    while T:
        xi = S.random_element()
        S = S.difference(Set([xi]))
        q = [valuation((a^xi*b)%p,k) for k in B] 
        if prod([k^valuation((a^xi*b)%p,k) for k in B]) == (a^xi*b)%p:
            DL = (sum(q[k]*LOGs[k] for k in [0..len(q)-1]) -xi)%(p-1)
            return DL
        
    

In [66]:
Index_calculas(3,37,1217,[2,3,5,7,11])

588

In [67]:
Index_calculas(2,37,131,[2,3,5,7])

41

### Pohlig-Hellman algorithm

Suppose we have a group $G = <g>$ such that ord(g) = N = $\prod_{k =1}^{n}{p_K}^{e_k}$ for some primes $p_1,p_2,\cdots,p_n$ and positive integers $e_1,e_2,\cdots,e_n$. Our goal is to find an $x$ for which $g^x = h$. Instead of solving this , first we try to compute $x $(mod ${p_i}^{e_i}$, then apply chinese remainder theorem to obtain $x$ mod($N$). So, we need to solve DLP in
$C_{{p_{i}}^{e_i}}$, where $C_{{p_{i}}^{e_i}}$ is the  cyclic group of order ${p_i}^{e_i}$. So, for all prime divisors of N, we get a system of congruences 

$x \equiv x_1$mod(${p_1}^{e_1})$

$x \equiv x_2$mod(${p_2}^{e_2})$

$\vdots$

$x \equiv x_n$mod(${p_n}^{e_n})$

By applying CRT, we get the solution for the system of linear congruence equations  $x$ mod(${p_1}^{e_1}{p_2}^{e_2}\cdots{p_n}^{e_n}) = N$.

So, we need to only solve the DLP in $C_{{p}^{e}}$. 

Let we want to determine $x$ for which $g^x = h$ in $C_{{p}^{e}}$. Let $x = b_0 + b_1 p + b_2 p^2 + \cdots + b_{e-1}p^{e-1}$.

In recurrent way,,let $x \equiv b_0 + b_1 p + b_2 p^2 + \cdots + b_{k-1}p^{k-1}$mod ($p^k)$ for $0 \le k \le e-1$. Then $x \equiv x_k + b_k p^k$mod($p^{k+1})$, that means  $h = g^{x_k}(g^{p^k})^{b_k}$.

Let $h_k = hg^{-x_k}$ and $g_k = g^{p^k}$. Hence, the reduced DLP is $h_k = {g_k}^{b_k}$.

In [68]:
def Pohlig_hellman(g,h,p):
    F = GF(p) # group of finite field of order p
    g1 = F(g) # return element g1 mod p
    h1 = F(h) # return element h1 mod p
    N = p-1
    qi = [r^N.valuation(r) for r in prime_divisors(N)]
    lqi = len(qi)
    Nqi = [N/q for q in qi]
    gi = [g1^r for r in Nqi]
    hi = [h1^r for r in Nqi]
    xi = [discrete_log(hi[i],gi[i]) for i in range(lqi)]
    x = CRT(xi,qi)
    return x

In [69]:
Pohlig_hellman(2,33,next_prime(12344))

9801